docs/rendering_iterables.md

```diff
@@ -0,0 +1,241 @@
+            lambda color, index: rx.box(rx.text(index), bg=color)
+        ),
+        columns=[2, 4, 6],
+    )
+
+```
+
+```python eval
+docdemo_from(ForeachIndexState, component=enumerate_foreach)
+```
+
+
+
+## Dictionary
+
+We can iterate through a `dict` data structure using a `foreach`. When the dict is passed through to the function that renders each item, it is presented as a list of key-value pairs `[("sky", "blue"), ("balloon", "red"), ("grass", "green")]`.
+
+```python exec
+class SimpleDictForeachState(rx.State):
+    color_chart: dict[str, str] = {
+        "sky": "blue",
+        "balloon": "red",
+        "grass": "green",
+    }
+
+
+def display_color(color: list):
+    # color is presented as a list key-value pairs [("sky", "blue"), ("balloon", "red"), ("grass", "green")]
+    return rx.box(rx.text(color[0]), bg=color[1], padding_x="1.5em")
+
+
+def dict_foreach():
+    return rx.responsive_grid(
+        rx.foreach(
+            SimpleDictForeachState.color_chart,
+            display_color,
+        ),
+        columns=[2, 4, 6],
+    )
+
+```
+
+
+```python eval
+docdemo_from(SimpleDictForeachState, display_color, component=dict_foreach)
+```
+
+
+
+
+## Nested examples
+
+`rx.foreach` can be used with nested state vars. Here we use nested `foreach` components to render the nested state vars. The `rx.foreach(project["technologies"], get_badge)` inside of the `project_item` function, renders the `dict` values which are of type `list`. The `rx.box(rx.foreach(NestedState.projects, project_item))` inside of the `projects_example` function renders each `dict` inside of the overall state var `projects`.
+
+```python exec
+class NestedState(rx.State):
+    projects: list[dict[str, list]] = [
+        {
+            "technologies": ["Next.js", "Prisma", "Tailwind", "Google Cloud", "Docker", "MySQL"]
+        },
+        {
+            "technologies": ["Python", "Flask", "Google Cloud", "Docker"]
+        }
+    ]
+
+def get_badge(technology: str) -> rx.Component:
+    return rx.badge(technology, variant="subtle", color_scheme="green")
+
+def project_item(project: dict) -> rx.Component:
+    return rx.box(
+        rx.hstack(            
+            rx.foreach(project["technologies"], get_badge)
+        ),
+    )
+
+def projects_example() -> rx.Component:
+    return rx.box(rx.foreach(NestedState.projects, project_item))
+```
+
+```python eval
+docdemo_from(NestedState, get_badge, project_item, component=projects_example)
+```
+
+
+If you want an example where not all of the values in the dict are the same type then check out the example on [var operations using foreach]({var_operations.path}).
+
+Here is a further example of how to use `foreach` with a nested data structure.
+
+```python exec
+class NestedDictForeachState(rx.State):
+    color_chart: dict[str, list[str]] = {
+        "purple": ["red", "blue"],
+        "orange": ["yellow", "red"],
+        "green": ["blue", "yellow"],
+    }
+
+
+def display_colors(color: list):
```

Comment:
i wonder if this would look better with a fuller annotation, like `list[str, list[str]]`

Reply:
updated to include this